internal/unionstore/memdb_art.go

```diff
@@ -165,3 +196,153 @@ func (db *artDBWithContext) SnapshotIterReverse(upper, lower []byte) Iterator {
 func (db *artDBWithContext) SnapshotGetter() Getter {
 	return db.ART.SnapshotGetter()
 }
+
+type snapshotBatchedIter struct {
+	db                    *artDBWithContext
+	snapshotTruncateSeqNo int
+	lower                 []byte
+	upper                 []byte
+	reverse               bool
+
+	// current batch
+	keys      [][]byte
+	values    [][]byte
+	pos       int
+	batchSize int
+	nextKey   []byte
+}
+
+func (db *artDBWithContext) BatchedSnapshotIter(lower, upper []byte, reverse bool) Iterator {
+	if len(db.Stages()) == 0 {
+		logutil.BgLogger().Error("should not use BatchedSnapshotIter for a memdb without any staging buffer")
+	}
+	iter := &snapshotBatchedIter{
+		db:                    db,
+		snapshotTruncateSeqNo: db.SnapshotSeqNo,
+		lower:                 lower,
+		upper:                 upper,
+		reverse:               reverse,
+		batchSize:             32,
+	}
+
+	iter.fillBatch()
+	return iter
+}
+
+func (it *snapshotBatchedIter) fillBatch() error {
+	if it.snapshotTruncateSeqNo != it.db.SnapshotSeqNo {
+		return errors.New(fmt.Sprintf("invalid iter: truncation happened, iter's=%d, db's=%d",
+			it.snapshotTruncateSeqNo, it.db.SnapshotSeqNo))
+	}
+
+	it.db.RLock()
+	defer it.db.RUnlock()
+
+	if it.keys == nil || it.values == nil || cap(it.keys) < it.batchSize || cap(it.values) < it.batchSize {
+		it.keys = make([][]byte, 0, it.batchSize)
+		it.values = make([][]byte, 0, it.batchSize)
+	} else {
+		it.keys = it.keys[:0]
+		it.values = it.values[:0]
+	}
+
+	var snapshotIter Iterator
+	if it.reverse {
+		searchUpper := it.upper
+		if it.nextKey != nil {
+			searchUpper = it.nextKey
+		}
+		snapshotIter = it.db.SnapshotIterReverse(searchUpper, it.lower)
+	} else {
+		searchLower := it.lower
+		if it.nextKey != nil {
+			searchLower = it.nextKey
+		}
+		snapshotIter = it.db.SnapshotIter(searchLower, it.upper)
+	}
+	defer snapshotIter.Close()
```

Comment:
Perhaps we need to check if the newly created `snapshotIter` contains the same snapshot compared with the previous one, so the whole `snapshotBatchedIter` has a consistent snapshot semantics.